IllegalArgumentException: Committee information must be derived from a state no older than the previous epoch

I've just upgraded to teku/v0.12.4-dev-4855f89c/linux-x86_64/-privatebuild-openjdk64bitservervm-java-13 and have started seeing this message very frequently (on average every 90 seconds).

This is the same message as reported in #2570. However I am opening a new issue as I _am_ producing attestations and getting them into blocks. I don't think this message relates relates to that issue.

```
12:40:12.244 INFO  - ^[[37mSlot Event  *** Slot: 64700, Block: 637850..fb65, Epoch: 2021, Finalized checkpoint: 2019, Finalized root: 90a6f8..b678, Peers: 37^[[0m
12:40:16.226 INFO  - ^[[34mValidator   *** Published aggregate    Count: 1, Slot: 64700, Root: 29695d..db7f^[[0m
12:40:24.244 INFO  - ^[[37mSlot Event  *** Slot: 64701, Block:    ... empty, Epoch: 2021, Finalized checkpoint: 2019, Finalized root: 90a6f8..b678, Peers: 39^[[0m
12:40:32.316 INFO  - ^[[34mValidator   *** Published attestation  Count: 2, Slot: 64702, Root: 637850..fb65^[[0m
12:40:35.421 WARN  - Encountered exception while processing message for topic /eth2/e7a75d5a/beacon_attestation_17/ssz_snappy
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Committee information must be derived from a state no older than the previous epoch. State at slot 64639 is older than cutoff slot 64640
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:683) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658) ~[?:?]
        at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2165) [?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.thenApply(SafeFuture.java:320) ~[teku-infrastructure-async-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.validation.AttestationValidator.singleOrAggregateAttestationChecks(AttestationValidator.java:125) ~[teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.validation.AttestationValidator.validate(AttestationValidator.java:69) ~[teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.SingleAttestationTopicHandler.validateData(SingleAttestationTopicHandler.java:100) ~[teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.SingleAttestationTopicHandler.validateData(SingleAttestationTopicHandler.java:27) ~[teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.Eth2TopicHandler.lambda$handleMessage$2(Eth2TopicHandler.java:39) ~[teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1183) [?:?]
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2306) [?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.thenCompose(SafeFuture.java:364) [teku-infrastructure-async-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.Eth2TopicHandler.handleMessage(Eth2TopicHandler.java:37) [teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.p2p.libp2p.gossip.GossipHandler.apply(GossipHandler.java:77) [teku-networking-p2p-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.p2p.libp2p.gossip.GossipHandler.apply(GossipHandler.java:34) [teku-networking-p2p-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at io.libp2p.pubsub.PubsubApiImpl.onNewMessage(PubsubApiImpl.kt:79) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.pubsub.PubsubApiImpl.access$onNewMessage(PubsubApiImpl.kt:22) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.pubsub.PubsubApiImpl$1.invoke(PubsubApiImpl.kt:63) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.pubsub.PubsubApiImpl$1.invoke(PubsubApiImpl.kt:22) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.pubsub.AbstractRouter.onInbound(AbstractRouter.kt:212) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.etc.util.P2PService.streamInbound(P2PService.kt:171) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.etc.util.P2PService$StreamHandler$channelRead$1.invoke(P2PService.kt:51) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.etc.util.P2PService$StreamHandler$channelRead$1.invoke(P2PService.kt:36) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at io.libp2p.etc.util.P2PService$runOnEventThread$1.run(P2PService.kt:231) [jvm-libp2p-minimal-0.5.5-RELEASE.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: java.lang.IllegalArgumentException: Committee information must be derived from a state no older than the previous epoch. State at slot 64639 is older than cutoff slot 64640
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:441) ~[guava-29.0-jre.jar:?]
        at tech.pegasys.teku.datastructures.util.CommitteeUtil.validateStateForCommitteeQuery(CommitteeUtil.java:252) ~[teku-ethereum-datastructures-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.datastructures.util.CommitteeUtil.get_beacon_committee(CommitteeUtil.java:229) ~[teku-ethereum-datastructures-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at tech.pegasys.teku.networking.eth2.gossip.topics.validation.AttestationValidator.lambda$singleOrAggregateAttestationChecks$1(AttestationValidator.java:142) ~[teku-networking-eth2-0.12.4-SNAPSHOT.jar:0.12.4-dev-4855f89c]
        at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680) ~[?:?]
        ... 29 more

--- a/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/AggregateGenerator.java
+++ b/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/AggregateGenerator.java
@@ -13,13 +13,15 @@
 
 package tech.pegasys.teku.core;
 
-import static org.assertj.core.util.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkNotNull;
 
 import java.util.List;
 import java.util.NoSuchElementException;
 import java.util.Optional;
 import tech.pegasys.teku.bls.BLSKeyPair;
 import tech.pegasys.teku.bls.BLSSignature;
+import tech.pegasys.teku.core.exceptions.EpochProcessingException;
+import tech.pegasys.teku.core.exceptions.SlotProcessingException;
 import tech.pegasys.teku.core.signatures.LocalMessageSignerService;
 import tech.pegasys.teku.core.signatures.Signer;
 import tech.pegasys.teku.core.signatures.UnprotectedSigner;
@@ -50,12 +52,12 @@ public class AggregateGenerator {
   }
 
   public SignedAggregateAndProof validAggregateAndProof(final BeaconBlockAndState blockAndState) {
-    return generator().blockAndState(blockAndState).generate();
+    return generator().blockAndState(blockAndState, blockAndState.getSlot()).generate();
   }
 
   public SignedAggregateAndProof validAggregateAndProof(
       final BeaconBlockAndState blockAndState, final UInt64 slot) {
-    return generator().blockAndState(blockAndState).slot(slot).generate();
+    return generator().blockAndState(blockAndState, slot).generate();
   }
 
   private Signer getSignerForValidatorIndex(final int validatorIndex) {
@@ -65,20 +67,24 @@ public class AggregateGenerator {
   }
 
   public class Generator {
+    // The latest block being attested to
     private BeaconBlockAndState blockAndState;
+    // The latest state processed through to the current slot
+    private BeaconState stateAtSlot;
+    private UInt64 slot;
     private Optional<UInt64> aggregatorIndex = Optional.empty();
-    private Optional<UInt64> slot = Optional.empty();
     private Optional<Attestation> aggregate = Optional.empty();
     private Optional<BLSSignature> selectionProof = Optional.empty();
     private Optional<UInt64> committeeIndex = Optional.empty();
 
     public Generator blockAndState(final BeaconBlockAndState blockAndState) {
-      this.blockAndState = blockAndState;
-      return this;
+      return blockAndState(blockAndState, blockAndState.getSlot());
     }
 
-    public Generator slot(final UInt64 slot) {
-      this.slot = Optional.of(slot);
+    public Generator blockAndState(final BeaconBlockAndState blockAndState, final UInt64 slot) {
+      this.slot = slot;
+      this.blockAndState = blockAndState;
+      this.stateAtSlot = generateHeadState(blockAndState.getState(), slot);
       return this;
     }
 
@@ -103,8 +109,8 @@ public class AggregateGenerator {
     }
 
     public SignedAggregateAndProof generate() {
-      checkNotNull(blockAndState, "Missing block and state");
-      final UInt64 slot = this.slot.orElseGet(blockAndState::getSlot);
+      checkNotNull(blockAndState, "Missing block");
+      checkNotNull(slot, "Missing slot");
       final Attestation aggregate = this.aggregate.orElseGet(() -> createAttestation(slot));
 
       return this.aggregatorIndex
@@ -114,13 +120,12 @@ public class AggregateGenerator {
 
     private SignedAggregateAndProof generateWithAnyValidAggregatorIndex(
         final Attestation aggregate) {
-      final BeaconState state = blockAndState.getState();
       final List<Integer> beaconCommittee =
           CommitteeUtil.get_beacon_committee(
-              state, aggregate.getData().getSlot(), aggregate.getData().getIndex());
+              stateAtSlot, aggregate.getData().getSlot(), aggregate.getData().getIndex());
       for (int validatorIndex : beaconCommittee) {
         final Optional<BLSSignature> maybeSelectionProof =
-            createValidSelectionProof(validatorIndex, state, aggregate);
+            createValidSelectionProof(validatorIndex, stateAtSlot, aggregate);
         if (maybeSelectionProof.isPresent()) {
           return generate(aggregate, UInt64.valueOf(validatorIndex), maybeSelectionProof.get());
         }
@@ -131,9 +136,8 @@ public class AggregateGenerator {
 
     private SignedAggregateAndProof generateWithFixedAggregatorIndex(
         final UInt64 slot, final Attestation aggregate, final UInt64 aggregatorIndex) {
-      final BeaconState state = blockAndState.getState();
       final BLSSignature validSelectionProof =
-          createSelectionProof(aggregatorIndex.intValue(), state, slot);
+          createSelectionProof(aggregatorIndex.intValue(), stateAtSlot, slot);
       return generate(aggregate, aggregatorIndex, validSelectionProof);
     }
 
@@ -144,8 +148,7 @@ public class AggregateGenerator {
       final BLSSignature selectionProof = this.selectionProof.orElse(validSelectionProof);
       final AggregateAndProof aggregateAndProof =
           new AggregateAndProof(aggregatorIndex, aggregate, selectionProof);
-      return createSignedAggregateAndProof(
-          aggregatorIndex, aggregateAndProof, blockAndState.getState());
+      return createSignedAggregateAndProof(aggregatorIndex, aggregateAndProof, stateAtSlot);
     }
 
     private Attestation createAttestation(final UInt64 slot) {
@@ -193,6 +196,19 @@ public class AggregateGenerator {
               .signAggregateAndProof(aggregateAndProof, state.getForkInfo())
               .join();
       return new SignedAggregateAndProof(aggregateAndProof, aggregateSignature);
+    }
+
+    private BeaconState generateHeadState(final BeaconState state, final UInt64 slot) {
+      if (state.getSlot().equals(slot)) {
+        return state;
+      }
+
+      StateTransition stateTransition = new StateTransition();
+      try {
+        return stateTransition.process_slots(state, slot);
+      } catch (EpochProcessingException | SlotProcessingException e) {
+        throw new IllegalStateException(e);
+      }
     }
   }
 }

--- a/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/AttestationGenerator.java
+++ b/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/AttestationGenerator.java
@@ -32,6 +32,8 @@ import java.util.stream.StreamSupport;
 import tech.pegasys.teku.bls.BLS;
 import tech.pegasys.teku.bls.BLSKeyPair;
 import tech.pegasys.teku.bls.BLSSignature;
+import tech.pegasys.teku.core.exceptions.EpochProcessingException;
+import tech.pegasys.teku.core.exceptions.SlotProcessingException;
 import tech.pegasys.teku.core.signatures.LocalMessageSignerService;
 import tech.pegasys.teku.core.signatures.UnprotectedSigner;
 import tech.pegasys.teku.datastructures.blocks.BeaconBlock;
@@ -200,8 +202,10 @@ public class AttestationGenerator {
    * assigned slot.
    */
   private static class AttestationIterator implements Iterator<Attestation> {
-    // The head block to attest to with its corresponding state
-    private final BeaconBlockAndState headBlockAndState;
+    // The latest block being attested to
+    private final BeaconBlock headBlock;
+    // The latest state processed through to the current slot
+    private final BeaconState headState;
     // The assigned slot to generate attestations for
     private final UInt64 assignedSlot;
     // The epoch containing the assigned slot
@@ -218,12 +222,26 @@ public class AttestationGenerator {
         final UInt64 assignedSlot,
         final List<BLSKeyPair> validatorKeys,
         final Function<Integer, BLSKeyPair> validatorKeySupplier) {
-      this.headBlockAndState = headBlockAndState;
+      this.headBlock = headBlockAndState.getBlock();
+      this.headState = generateHeadState(headBlockAndState.getState(), assignedSlot);
       this.validatorKeys = validatorKeys;
       this.assignedSlot = assignedSlot;
       this.assignedSlotEpoch = compute_epoch_at_slot(assignedSlot);
       this.validatorKeySupplier = validatorKeySupplier;
       generateNextAttestation();
+    }
+
+    private BeaconState generateHeadState(final BeaconState state, final UInt64 slot) {
+      if (state.getSlot().equals(slot)) {
+        return state;
+      }
+
+      StateTransition stateTransition = new StateTransition();
+      try {
+        return stateTransition.process_slots(state, slot);
+      } catch (EpochProcessingException | SlotProcessingException e) {
+        throw new IllegalStateException(e);
+      }
     }
 
     public static AttestationIterator create(
@@ -269,8 +287,6 @@ public class AttestationGenerator {
     private void generateNextAttestation() {
       nextAttestation = Optional.empty();
 
-      final BeaconState headState = headBlockAndState.getState();
-      final BeaconBlock headBlock = headBlockAndState.getBlock();
       int lastProcessedValidatorIndex = currentValidatorIndex;
       for (int validatorIndex = currentValidatorIndex;
           validatorIndex < validatorKeys.size();

--- a/ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/forkchoice/ReadOnlyStore.java
+++ b/ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/forkchoice/ReadOnlyStore.java
@@ -86,4 +86,6 @@ public interface ReadOnlyStore {
   SafeFuture<Optional<BeaconState>> retrieveBlockState(Bytes32 blockRoot);
 
   SafeFuture<Optional<BeaconState>> retrieveCheckpointState(Checkpoint checkpoint);
+
+  BeaconState getCheckpointState(final Checkpoint checkpoint, final BeaconState latestStateAtEpoch);
 }

--- a/ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/forkchoice/TestStoreImpl.java
+++ b/ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/forkchoice/TestStoreImpl.java
@@ -173,6 +173,15 @@ class TestStoreImpl implements MutableStore {
     return SafeFuture.completedFuture(getCheckpointState(checkpoint));
   }
 
+  @Override
+  public BeaconState getCheckpointState(
+      final Checkpoint checkpoint, final BeaconState latestStateAtEpoch) {
+    if (!latestStateAtEpoch.getSlot().equals(checkpoint.getEpochStartSlot())) {
+      throw new UnsupportedOperationException("Checkpoint state calculation not supported");
+    }
+    return latestStateAtEpoch;
+  }
+
   // Mutable methods
   @Override
   public void putBlockAndState(final SignedBeaconBlock block, final BeaconState state) {

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/AttestationValidator.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/AttestationValidator.java
@@ -15,6 +15,7 @@ package tech.pegasys.teku.networking.eth2.gossip.topics.validation;
 
 import static tech.pegasys.teku.datastructures.util.AttestationUtil.get_indexed_attestation;
 import static tech.pegasys.teku.datastructures.util.AttestationUtil.is_valid_indexed_attestation;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_epoch_at_slot;
 import static tech.pegasys.teku.datastructures.util.CommitteeUtil.computeSubnetForAttestation;
 import static tech.pegasys.teku.datastructures.util.CommitteeUtil.get_beacon_committee;
 import static tech.pegasys.teku.infrastructure.unsigned.UInt64.ONE;
@@ -31,10 +32,13 @@ import java.util.List;
 import java.util.Objects;
 import java.util.OptionalInt;
 import java.util.Set;
+import org.apache.tuweni.bytes.Bytes32;
 import tech.pegasys.teku.datastructures.attestation.ValidateableAttestation;
 import tech.pegasys.teku.datastructures.operations.Attestation;
 import tech.pegasys.teku.datastructures.operations.IndexedAttestation;
 import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.datastructures.state.Checkpoint;
+import tech.pegasys.teku.datastructures.util.CommitteeUtil;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.storage.client.RecentChainData;
@@ -128,7 +132,7 @@ public class AttestationValidator {
                 return SAVE_FOR_FUTURE;
               }
 
-              final BeaconState state = maybeState.get();
+              final BeaconState state = resolveStateForAttestation(attestation, maybeState.get());
 
               // The attestation's committee index (attestation.data.index) is for the correct
               // subnet.
@@ -153,6 +157,33 @@ public class AttestationValidator {
               }
               return ACCEPT;
             });
+  }
+
+  /**
+   * Committee information is only guaranteed to be stable up to 1 epoch ahead, if block attested to
+   * is too old, we need to roll the corresponding state forward to process the attestation
+   *
+   * @param attestation The attestation to be processed
+   * @param blockState The state corresponding to the block being attested to
+   * @return The state to use for validation of this attestation
+   */
+  public BeaconState resolveStateForAttestation(
+      final Attestation attestation, final BeaconState blockState) {
+    final Bytes32 blockRoot = attestation.getData().getBeacon_block_root();
+    final Checkpoint targetEpoch = attestation.getData().getTarget();
+    final UInt64 earliestSlot =
+        CommitteeUtil.getEarliestQueryableSlotForTargetEpoch(targetEpoch.getEpoch());
+    final UInt64 earliestEpoch = compute_epoch_at_slot(earliestSlot);
+
+    final BeaconState state;
+    if (blockState.getSlot().isLessThan(earliestSlot)) {
+      final Checkpoint checkpoint = new Checkpoint(earliestEpoch, blockRoot);
+      state = recentChainData.getStore().getCheckpointState(checkpoint, blockState);
+    } else {
+      state = blockState;
+    }
+
+    return state;
   }
 
   private ValidatorAndTargetEpoch getValidatorAndTargetEpoch(final Attestation attestation) {

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidator.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidator.java
@@ -108,7 +108,11 @@ public class SignedAggregateAndProofValidator {
                         if (maybeState.isEmpty()) {
                           return SAVE_FOR_FUTURE;
                         }
-                        final BeaconState state = maybeState.get();
+
+                        final BeaconState state =
+                            attestationValidator.resolveStateForAttestation(
+                                aggregate, maybeState.get());
+
                         final Optional<BLSPublicKey> aggregatorPublicKey =
                             ValidatorsUtil.getValidatorPubKey(state, aggregateAndProof.getIndex());
                         if (aggregatorPublicKey.isEmpty()) {

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/AttestationValidatorTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/AttestationValidatorTest.java
@@ -104,6 +104,17 @@ class AttestationValidatorTest {
   }
 
   @Test
+  public void shouldReturnValidForValidAttestation_whenManyBlocksHaveBeenSkipped() {
+    final BeaconBlockAndState head = recentChainData.getHeadBlockAndState().orElseThrow();
+    final UInt64 currentSlot = head.getSlot().plus(SLOTS_PER_EPOCH * 3);
+    storageSystem.chainUpdater().setCurrentSlot(currentSlot);
+
+    final Attestation attestation =
+        attestationGenerator.validAttestation(head, head.getSlot().plus(SLOTS_PER_EPOCH * 3));
+    assertThat(validate(attestation)).isEqualTo(ACCEPT);
+  }
+
+  @Test
   public void shouldRejectAttestationWithIncorrectAggregateBitsSize() {
     final Attestation attestation =
         attestationGenerator.validAttestation(recentChainData.getHeadBlockAndState().orElseThrow());

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidatorTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidatorTest.java
@@ -29,6 +29,7 @@ import static tech.pegasys.teku.networking.eth2.gossip.topics.validation.Interna
 import static tech.pegasys.teku.networking.eth2.gossip.topics.validation.InternalValidationResult.IGNORE;
 import static tech.pegasys.teku.networking.eth2.gossip.topics.validation.InternalValidationResult.REJECT;
 import static tech.pegasys.teku.networking.eth2.gossip.topics.validation.InternalValidationResult.SAVE_FOR_FUTURE;
+import static tech.pegasys.teku.util.config.Constants.SLOTS_PER_EPOCH;
 
 import java.util.List;
 import java.util.OptionalInt;
@@ -112,6 +113,7 @@ class SignedAggregateAndProofValidatorTest {
   private final SignedAggregateAndProofValidator validator =
       new SignedAggregateAndProofValidator(recentChainData, attestationValidator);
   private SignedBlockAndState bestBlock;
+  private SignedBlockAndState genesis;
 
   @BeforeAll
   public static void init() {
@@ -125,7 +127,7 @@ class SignedAggregateAndProofValidatorTest {
 
   @BeforeEach
   public void setUp() {
-    chainUpdater.initializeGenesis(false);
+    genesis = chainUpdater.initializeGenesis(false);
     bestBlock = chainUpdater.addNewBestBlock();
 
     final AttestationValidator realAttestationValidator = new AttestationValidator(recentChainData);
@@ -140,6 +142,19 @@ class SignedAggregateAndProofValidatorTest {
   public void shouldReturnValidForValidAggregate() {
     final BeaconBlockAndState chainHead = recentChainData.getHeadBlockAndState().orElseThrow();
     final SignedAggregateAndProof aggregate = generator.validAggregateAndProof(chainHead);
+    whenAttestationIsValid(aggregate);
+    assertThat(validator.validate(ValidateableAttestation.fromSignedAggregate(aggregate)))
+        .isCompletedWithValue(ACCEPT);
+  }
+
+  @Test
+  public void shouldReturnValidForValidAggregate_whenManyBlocksHaveBeenSkipped() {
+    final BeaconBlockAndState chainHead = recentChainData.getHeadBlockAndState().orElseThrow();
+    final UInt64 currentSlot = chainHead.getSlot().plus(SLOTS_PER_EPOCH * 3);
+    storageSystem.chainUpdater().setCurrentSlot(currentSlot);
+
+    final SignedAggregateAndProof aggregate =
+        generator.validAggregateAndProof(chainHead, currentSlot);
     whenAttestationIsValid(aggregate);
     assertThat(validator.validate(ValidateableAttestation.fromSignedAggregate(aggregate)))
         .isCompletedWithValue(ACCEPT);
@@ -336,8 +351,7 @@ class SignedAggregateAndProofValidatorTest {
     final SignedAggregateAndProof aggregateAndProof1 =
         generator
             .generator()
-            .blockAndState(chainHead)
-            .slot(ZERO)
+            .blockAndState(genesis.toUnsigned())
             .aggregatorIndex(aggregatorIndex)
             .generate();
 
@@ -346,9 +360,8 @@ class SignedAggregateAndProofValidatorTest {
     final SignedAggregateAndProof aggregateAndProof2 =
         generator
             .generator()
-            .blockAndState(chainHead)
+            .blockAndState(chainHead, epochOneCommitteeAssignment.getSlot())
             .committeeIndex(epochOneCommitteeAssignment.getCommitteeIndex())
-            .slot(epochOneCommitteeAssignment.getSlot())
             .aggregatorIndex(aggregatorIndex)
             .generate();
     whenAttestationIsValid(aggregateAndProof1);
@@ -375,10 +388,9 @@ class SignedAggregateAndProofValidatorTest {
     final SignedAggregateAndProof aggregate =
         generator
             .generator()
-            .blockAndState(chainHead)
+            .blockAndState(chainHead, committeeAssignment.getSlot())
             .aggregatorIndex(UInt64.valueOf(aggregatorIndex))
             .committeeIndex(committeeAssignment.getCommitteeIndex())
-            .slot(committeeAssignment.getSlot())
             .generate();
     whenAttestationIsValid(aggregate);
     // Sanity check

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidatorTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/SignedAggregateAndProofValidatorTest.java
@@ -15,6 +15,7 @@ package tech.pegasys.teku.networking.eth2.gossip.topics.validation;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.fail;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
@@ -126,6 +127,13 @@ class SignedAggregateAndProofValidatorTest {
   public void setUp() {
     chainUpdater.initializeGenesis(false);
     bestBlock = chainUpdater.addNewBestBlock();
+
+    final AttestationValidator realAttestationValidator = new AttestationValidator(recentChainData);
+    when(attestationValidator.resolveStateForAttestation(any(), any()))
+        .thenAnswer(
+            i ->
+                realAttestationValidator.resolveStateForAttestation(
+                    i.getArgument(0), i.getArgument(1)));
   }
 
   @Test

--- a/storage/src/main/java/tech/pegasys/teku/storage/store/Store.java
+++ b/storage/src/main/java/tech/pegasys/teku/storage/store/Store.java
@@ -451,6 +451,17 @@ class Store implements UpdatableStore {
                     }));
   }
 
+  @Override
+  public BeaconState getCheckpointState(
+      final Checkpoint checkpoint, final BeaconState latestStateAtEpoch) {
+    if (latestStateAtEpoch.getSlot().isGreaterThan(checkpoint.getEpochStartSlot())) {
+      throw new IllegalArgumentException("Latest state must be at or prior to checkpoint slot");
+    }
+    final BeaconState checkpointState = regenerateCheckpointState(checkpoint, latestStateAtEpoch);
+    putCheckpointState(checkpoint, checkpointState);
+    return checkpointState;
+  }
+
   private Optional<BeaconState> getCheckpointStateIfAvailable(final Checkpoint checkpoint) {
     readLock.lock();
     try {
@@ -837,6 +848,12 @@ class Store implements UpdatableStore {
             Optional.of(regenerateCheckpointState(checkpoint, inMemoryCheckpointBlockState)));
       }
       return Store.this.retrieveCheckpointState(checkpoint);
+    }
+
+    @Override
+    public BeaconState getCheckpointState(
+        final Checkpoint checkpoint, final BeaconState latestStateAtEpoch) {
+      return Store.this.getCheckpointState(checkpoint, latestStateAtEpoch);
     }
 
     @Override

--- a/storage/src/test/java/tech/pegasys/teku/storage/store/StoreTest.java
+++ b/storage/src/test/java/tech/pegasys/teku/storage/store/StoreTest.java
@@ -137,6 +137,56 @@ class StoreTest extends AbstractStoreTest {
   }
 
   @Test
+  public void getCheckpointState_forGenesis() {
+    final SignedBlockAndState genesisBlockAndState = chainBuilder.generateGenesis();
+    final AnchorPoint genesis = AnchorPoint.fromGenesisState(genesisBlockAndState.getState());
+    final BlockProvider blockProvider = blockProviderFromChainBuilder();
+
+    final UpdatableStore store =
+        StoreBuilder.buildForkChoiceStore(new StubMetricsSystem(), blockProvider, genesis).join();
+    final Checkpoint checkpoint = new Checkpoint(UInt64.ZERO, genesisBlockAndState.getRoot());
+
+    final BeaconState baseState = genesisBlockAndState.getState();
+    final BeaconState result = store.getCheckpointState(checkpoint, baseState);
+    assertThat(result).isEqualTo(baseState);
+  }
+
+  @Test
+  public void getCheckpointState_forEpochPastGenesis() {
+    final SignedBlockAndState genesisBlockAndState = chainBuilder.generateGenesis();
+    final AnchorPoint genesis = AnchorPoint.fromGenesisState(genesisBlockAndState.getState());
+    final BlockProvider blockProvider = blockProviderFromChainBuilder();
+
+    final UpdatableStore store =
+        StoreBuilder.buildForkChoiceStore(new StubMetricsSystem(), blockProvider, genesis).join();
+    final Checkpoint checkpoint = new Checkpoint(UInt64.ONE, genesisBlockAndState.getRoot());
+
+    final BeaconState baseState = genesisBlockAndState.getState();
+    final BeaconState result = store.getCheckpointState(checkpoint, baseState);
+    assertThat(result.getSlot()).isGreaterThan(baseState.getSlot());
+    assertThat(result.getSlot()).isEqualTo(checkpoint.getEpochStartSlot());
+    assertThat(result.getLatest_block_header().hash_tree_root()).isEqualTo(checkpoint.getRoot());
+  }
+
+  @Test
+  public void getCheckpointState_invalidState() {
+    final SignedBlockAndState genesisBlockAndState = chainBuilder.generateGenesis();
+    final AnchorPoint genesis = AnchorPoint.fromGenesisState(genesisBlockAndState.getState());
+    final BlockProvider blockProvider = blockProviderFromChainBuilder();
+
+    final UpdatableStore store =
+        StoreBuilder.buildForkChoiceStore(new StubMetricsSystem(), blockProvider, genesis).join();
+    final SignedBlockAndState futureBlockAndState =
+        chainBuilder.generateBlockAtSlot(compute_start_slot_at_epoch(UInt64.valueOf(2)));
+
+    final Checkpoint checkpoint = new Checkpoint(UInt64.ONE, futureBlockAndState.getRoot());
+
+    assertThatThrownBy(() -> store.getCheckpointState(checkpoint, futureBlockAndState.getState()))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Latest state must be at or prior to checkpoint slot");
+  }
+
+  @Test
   public void
       retrieveCheckpointState_shouldThrowInvalidCheckpointExceptionWhenEpochBeforeBlockRoot()
           throws Exception {

